src/lib.rs

```diff
@@ -439,6 +465,145 @@ impl DstLayout {
+        //   fields, is the smaller of the specified alignment and the alignment
+        //   of the field's type.
+        let field_align = min(field._align, max_align);
+
+        // The struct's alignment is the maximum of its previous alignment and
+        // `field_align`.
+        let align = max(self._align, field_align);
+
+        let size_info = match self._size_info {
+            // If the layout is already a DST, we panic; DSTs cannot be extended
+            // with additional fields.
+            SizeInfo::SliceDst(..) => panic!("Cannot extend a DST with additional fields."),
+
+            SizeInfo::Sized { _size: preceeding_size } => {
+                // Compute the minimum amount of inter-field padding needed to
+                // satisfy the field's alignment, and offset of the trailing
+                // field. [1]
+                //
+                // [1] Per https://doc.rust-lang.org/reference/type-layout.html#the-alignment-modifiers:
+                //
+                //   Inter-field padding is guaranteed to be the minimum
+                //   required in order to satisfy each field's (possibly
+                //   altered) alignment.
+                let padding = _padding_needed_for(preceeding_size, field_align);
+
+                // This will not panic (and is proven to not panic, with Kani)
+                // if the layout components can correspond to a leading layout
+                // fragment of a valid Rust type, but may panic otherwise (e.g.,
+                // combining or aligning the components would create a size
+                // exceeding `isize::MAX`).
+                let offset = match preceeding_size.checked_add(padding) {
+                    Some(offset) => offset,
+                    None => panic!("Adding padding to `self`'s size overflows `usize`."),
+                };
+
+                match field._size_info {
+                    SizeInfo::Sized { _size: field_size } => {
+                        // If the trailing field is sized, the resulting layout
+                        // will be sized. Its size will be the sum of the
+                        // preceeding layout, the size of the new field, and the
+                        // size of inter-field padding between the two.
+                        //
+                        // This will not panic (and is proven with Kani to not
+                        // panic) if the layout components can correspond to a
+                        // leading layout fragment of a valid Rust type, but may
+                        // panic otherwise (e.g., combining or aligning the
+                        // components would create a size exceeding
+                        // `isize::MAX`).
+                        let size = match offset.checked_add(field_size) {
+                            Some(size) => size,
+                            None => panic!("`field` cannot be appended without the total size overflowing `isize::MAX`"),
```

Comment:
This case is only triggered if the total overflows `usize`, not `isize`, right?

This raises an interesting question: Should we be doing more to guarantee that we don't overflow `isize`? Maybe validate using Kani that the output size fits in `isize`?

Reply:
I've fixed the panic message.

> Should we be doing more to guarantee that we don't overflow `isize`?

Maybe, though I'd argue doing so doesn't buy us much. We currently prove with kani that across `Layout::extend`'s domain, `DstLayout::extend` behaves analogously (among other correctness properties that guard against the possibility of a bug arising in `Layout::extend`).

We prove this by constructing `Layout` analogs to the `DstLayout` constituents, combining them with `Layout::extend`, and then verifying that the resulting `Layout` composite matches the `DstLayout` composite; e.g.:
```rust
if let Ok((actual_composite, actual_offset)) = base_analog.extend(field_analog)
{
    assert_eq!(actual_offset, offset);
    assert_eq!(actual_composite.size(), composite_size);
    assert_eq!(actual_composite.align(), composite._align.get());
} else {
    // An error here reflects that composite of `base`
    // and `field` cannot correspond to a real Rust type
    // fragment, because such a fragment would violate
    // the basic invariants of a valid Rust layout. At
    // the time of writing, `DstLayout` is a little more
    // permissive than `Layout`, so we don't assert
    // anything in this branch (e.g., unreachability).
}
```

This is enough to support the *Safety* section in `DstLayout::extend`'s documentation (and our use-case for this method):

> If a (potentially hypothetical) valid `repr(C)` Rust type begins with fields whose layout are `self`, and those fields are immediately followed by a field whose layout is `field`, then unsafe code may rely on `self.extend(field, repr_packed)` producing a layout that correctly encompasses those two components.
>
> We make no guarantees to the behavior of this method if these fragments cannot appear in a valid Rust type (e.g., the concatenation of the layouts would lead to a size larger than `isize::MAX`).

We *could* establish safety invariants on `DstLayout`'s fields (e.g., alignment is valid) and rigorously enforce those conditions, but we would need to weigh that against the maintenance costs of the potential for extra unsafety/unsoundness.

Given our currently limited use of `DstLayout` — only describing real Rust types — I think that we should punt on adding these safety conditions until we actually need them.

Reply:
Ack.